keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1036 @@
+-->
+
+This KEP is scoped to CRDs, but if [KEP-3937](https://github.com/kubernetes/enhancements/pull/3938) is implemented then we should expect it to also affect native type validation.
+
+
+#### Schema-Changing Tests May Need Refactoring
+
+This change could cause some previously valid tests of invalid schema changes to fail. Those tests will have to be updated by changing to use create, or by
+changing the invalid fields during an update.
+
+#### Detection of Breaking Schema Changes is Different
+
+Some users might have a process to detect breaking schema changes by applying
+their CRD and poking an existing resource. This KEP would make that process
+no longer reliable.
+
+##### Mitigations
+###### Use CRD-Schema-Checker
+
+[CRD-Schema-Checker](https://github.com/openshift/crd-schema-checker) may be
+used to lint the CRD schema for breaking changes before applying it to a cluster.
+
+###### Use New Objects
+
+Users may cerate new objects rather than updating old ones to test for breaking schema changes. This feature has no effect on the creation code path.
+
+### Risks and Mitigations
+
+Help Wanted
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The basic idea is to add a post-processing step to the `Validate(Update)` methods of the CRD strategy. Given a list of `StatusError`, we check the field path associated with the error, if it is unchanged from old to new then the error is changed into a warning.
+
+### Correlating Errors To Fields
+
+There are a few problems with using the FieldPath associated with the error
+to potentially ignore errors:
+
+1. Prepending item to atomic list.
+
+This case changes the index of all items, and the list is atomic, so it is not possible to ratchet individual items in an atomic list.
+
+2. Field Path containing list-type=map still contains a position index relative to the new object. 
+
+We will have to look up the map key in the new object to find the element's map key. Using the map key we can find the associated element in the old object.
+
+3. Field Path in Errors from CEL Rules may be a lie: https://github.com/kubernetes/kubernetes/pull/118041
+
+The field to which the CEL rule in its `x-kubernetes-validations` list must 
+contain all fields referred by the CEL rule; so the CEL error logic can have to 
+be modified to also include the rule's original location.
+
```

Comment:
Something to clarify:
```yaml
type: object
properties:
  a:
    type: string
  b:
    type: string
  c:
    type: string
  or: # this is new
  - properties:
       a:
          enum: ["foo"]
  - properties:
       b:
          enum: ["bar"]
```
This will result in a json path `.` for `{a:"x", b:"x",c:"1"} -> {a:"x", b:"x",c:"2"}`. And the value at `.` has obviously changed. But the value validation has been false before and is false after, the change in `c` does not matter.

Does this update pass?

Reply:
Yes, this update would pass. The new validations are associated with specific fields `.a` and `.b` rather than the root object, so the validator would be able to correlate the old and new values. Thank you for providing this edge case, I will add this and a few more to the doc

Reply:
I realized after posting that since "anyOf" requires "one or more" schemas to match, there will be an error on the main object from the `anyOf` field itself. I see now the question in this case is when evaluating the `anyOf`, do we consider the ignored errors or not? 

1. If the validator ignores the errors for unchanged fields internally, the update should pass since any raised errors were suppressed.
2. If the validator has no internal logic for this, and reports the error on the root field path then the update would fail since "c" was changed.

I am leaning towards Option 1 since it allows the user more flexibility, but I will have to think through and document its behavior for all of the nested value validations. Thank you for pointing this out.

Reply:
Added notes about these cases to the KEP.